specs/IFramePermissionRequested.md

```diff
@@ -0,0 +1,363 @@
+                                [this](ICoreWebView2Frame* sender,
+                                   ICoreWebView2PermissionRequestedEventArgs* args) -> HRESULT {
+                                    COREWEBVIEW2_PERMISSION_KIND kind =
+                                        COREWEBVIEW2_PERMISSION_KIND_UNKNOWN_PERMISSION;
+                                    BOOL userInitiated = FALSE;
+                                    wil::unique_cotaskmem_string uri;
+
+                                    CHECK_FAILURE(args->get_PermissionKind(&kind));
+                                    CHECK_FAILURE(args->get_IsUserInitiated(&userInitiated));
+                                    CHECK_FAILURE(args->get_Uri(&uri));
+
+                                    auto cachedKey = std::tuple<
+                                        std::wstring, COREWEBVIEW2_PERMISSION_KIND, BOOL>(
+                                        std::wstring(uri.get()), kind, userInitiated);
+
+                                    auto cachedPermission =
+                                        m_cachedPermissions.find(cachedKey);
+                                    if (cachedPermission != m_cachedPermissions.end())
+                                    {
+                                        bool allow = cachedPermission->second;
+                                        if (allow)
+                                        {
+                                            CHECK_FAILURE(args->put_State(
+                                                COREWEBVIEW2_PERMISSION_STATE_ALLOW));
+                                        }
+                                        else
+                                        {
+                                            CHECK_FAILURE(args->put_State(
+                                                COREWEBVIEW2_PERMISSION_STATE_DENY));
+                                        }
+
+                                        PutHandled(args);
+                                        return S_OK;
+                                    }
+
+                                    std::wstring message =
+                                        L"An iframe has requested device permission for ";
+                                    message += NameOfPermissionKind(kind);
+                                    message += L" to the website at ";
+                                    message += uri.get();
+                                    message += L"?\n\n";
+                                    message += L"Do you want to grant permission?\n";
+                                    message +=
+                                        (userInitiated
+                                             ? L"This request came from a user gesture."
+                                             : L"This request did not come from a user "
+                                               L"gesture.");
+
+                                    int response = MessageBox(
```

Comment:
We don't want to run a message loop during an event handler or async callback. You'll need to get a deferral and run this async. There should be examples in the sample app of this - I think we have a helper function that takes like a lambda to run later asynchronously on the UI thread.

Reply:
Added async call w deferral